src/azfilenfs-watchdog

```diff
@@ -0,0 +1,222 @@
+#!/bin/bash
+
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+#
+# How often does the watchdog look for unmounts and/or to track if stunnel processes are running for
+# nfs file endpoints.
+#
+MONITOR_INTERVAL_SECS=5
+
+#
+# Don't perform mountmap and iptables rule cleanup for unmounted filesystems.
+# This can be set if you want lazy umount to work.
+#
+AZNFS_SKIP_UNMOUNT_CLEANUP="${AZNFS_SKIP_UNMOUNT_CLEANUP:-0}"
+
+# Load common aznfs helpers.
+. /opt/microsoft/aznfs/common.sh
+AZ_FILES_MOUNTMAP="$OPTDIR/aznfs_files_mountmap"
+
+
+#
+# Delete entry from $AZ_FILES_MOUNTMAP.
+#
+ensure_az_files_mountmap_not_exist()
+{
+    (
+        flock -e 999
+
+        #
+        # If user wants to delete the entry only if AZ_FILES_MOUNTMAP has not changed since
+        # he looked up, honour that.
+        #
+        local ifmatch="$2"
+        if [ -n "$ifmatch" ]; then
+            local mtime=$(stat -c%Y $AZ_FILES_MOUNTMAP)
+            if [ "$mtime" != "$ifmatch" ]; then
+                eecho "[$1] Refusing to remove from ${AZ_FILES_MOUNTMAP} as $mtime != $ifmatch!"
+                return 1
+            fi
+        fi
+
+	chattr -f -i $AZ_FILES_MOUNTMAP
+        #
+        # We do this thing instead of inplace update by sed as that has a
+        # very bad side-effect of creating a new AZ_FILES_MOUNTMAP file. This breaks
+        # any locking that we dependent on the old file.
+        #
+        out=$(sed "\%^${1}$%d" $AZ_FILES_MOUNTMAP)
+        ret=$?
+        if [ $ret -eq 0 ]; then
+            #
+            # If this echo fails then AZ_FILES_MOUNTMAP could be truncated.
+            #
+            echo "$out" > $AZ_FILES_MOUNTMAP
+            ret=$?
+            out=
+            if [ $ret -ne 0 ]; then
+                eecho "*** [FATAL] AZ_FILES_MOUNTMAP may be in inconsistent state, contact Microsoft support ***"
+            fi
+        fi
+
+        if [ $ret -ne 0 ]; then
+            chattr -f +i $AZ_FILES_MOUNTMAP
+            eecho "[$1] failed to remove from ${AZ_FILES_MOUNTMAP}!"
+            return 1
+        fi
+        chattr -f +i $AZ_FILES_MOUNTMAP
+
+	# Return the mtime after our mods.
+        echo $(stat -c%Y $AZ_FILES_MOUNTMAP)
+    ) 999<$AZ_FILES_MOUNTMAP
+}
+
+
+vecho "Starting azfilenfs-watchdog..."
+
+if [ ! -f $AZ_FILES_MOUNTMAP ]; then
+	touch $AZ_FILES_MOUNTMAP
+	if [ $? -ne 0 ]; then
```

Comment:
what if you keep failing with permission issues on creating the mountmap file?

Reply:
we run watchdog script with root privileges, so there won't be permission issues on creating the mountmap file. If mountmap file creation fails due to some other reason, we log the error and exit. When customer tries to mount nfsv4 share, mount helper script checks if watchdog is running. If watchdog is not running, mount helper tries to restart watchdog 5 times, and then logs the error and exit if unable to restart the watchdog.